File: 数据/航道数据收集/track_centerline.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import interpolate
import os

# 读取Excel文件
def read_excel_file(file_path):
    try:
        # 尝试不同的sheet名称
        try:
            df = pd.read_excel(file_path, sheet_name='Sheet1')
        except:
            df = pd.read_excel(file_path)
        
        # 识别经纬度列
        lat_col = None
        lon_col = None
        
        # 检查列名
        for col in df.columns:
            col_lower = str(col).lower()
            if 'lat' in col_lower or '纬度' in col_lower:
                lat_col = col
            elif 'lon' in col_lower or '经度' in col_lower:
                lon_col = col
        
        # 如果没有找到，尝试根据数据特征识别
        if lat_col is None or lon_col is None:
            # 假设前几列中包含经纬度
            for i, col in enumerate(df.columns[:5]):
                data = df[col].dropna()
                if len(data) > 0:
                    # 纬度范围通常在-90到90之间
                    if (data.min() >= -90) and (data.max() <= 90):
                        lat_col = col
                    # 经度范围通常在-180到180之间
                    elif (data.min() >= -180) and (data.max() <= 180):
                        lon_col = col
        
        if lat_col is None or lon_col is None:
            print(f"无法识别文件 {file_path} 中的经纬度列")
            return None
        
        # 提取经纬度数据
        lons = df[lon_col].dropna().values
        lats = df[lat_col].dropna().values
        
        return lons, lats
    except Exception as e:
        print(f"读取文件 {file_path} 时出错: {e}")
        return None

# 计算轨迹中心线
def calculate_centerline(lons, lats):
    if len(lons) < 2:
        return lons, lats
    
    # 对轨迹进行插值，使点分布更均匀
    t = np.linspace(0, 1, len(lons))
    t_new = np.linspace(0, 1, max(100, len(lons)))
    
    # 使用样条插值
    lon_interp = interpolate.interp1d(t, lons, kind='cubic')
    lat_interp = interpolate.interp1d(t, lats, kind='cubic')
    
    lons_smooth = lon_interp(t_new)
    lats_smooth = lat_interp(t_new)
    
    return lons_smooth, lats_smooth

# 将轨迹分成上下两个通道
def split_into_channels(lons, lats):
    # 计算轨迹的方向
    dx = np.diff(lons)
    dy = np.diff(lats)
    angles = np.arctan2(dy, dx)
    
    # 计算平均方向
    mean_angle = np.mean(angles)
    
    # 计算垂直方向（向左和向右）
    perpendicular_angle1 = mean_angle + np.pi/2
    perpendicular_angle2 = mean_angle - np.pi/2
    
    # 计算偏移距离（可根据实际情况调整）
    offset = 0.0001  # 约10米
    
    # 计算上下通道的点
    lons_channel1 = lons + offset * np.cos(perpendicular_angle1)
    lats_channel1 = lats + offset * np.sin(perpendicular_angle1)
    
    lons_channel2 = lons + offset * np.cos(perpendicular_angle2)
    lats_channel2 = lats + offset * np.sin(perpendicular_angle2)
    
    return (lons_channel1, lats_channel1), (lons_channel2, lats_channel2)

# 可视化轨迹和中心线
def visualize_track(lons, lats, centerline, channel1, channel2, file_name):
    plt.figure(figsize=(12, 8))
    
    # 绘制原始轨迹
    plt.plot(lons, lats, 'b-', alpha=0.5, label='原始轨迹')
    
    # 绘制中心线
    plt.plot(centerline[0], centerline[1], 'r-', linewidth=2, label='轨迹中心线')
    
    # 绘制上下通道
    plt.plot(channel1[0], channel1[1], 'g-', linewidth=1.5, label='上通道')
    plt.plot(channel2[0], channel2[1], 'y-', linewidth=1.5, label='下通道')
    
    plt.xlabel('经度')
    plt.ylabel('纬度')
    plt.title(f'轨迹分析 - {file_name}')
    plt.legend()
    plt.grid(True)
    plt.axis('equal')
    
    # 保存图像
    output_dir = 'output'
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
    
    plt.savefig(os.path.join(output_dir, f'{os.path.splitext(file_name)[0]}_track.png'))
    plt.close()

# 处理单个文件
def process_file(file_path):
    print(f"处理文件: {file_path}")
    
    # 读取数据
    data = read_excel_file(file_path)
    if data is None:
        return
    
    lons, lats = data
    
    # 计算中心线
    centerline = calculate_centerline(lons, lats)
    
    # 分成上下通道
    channel1, channel2 = split_into_channels(centerline[0], centerline[1])
    
    # 可视化
    file_name = os.path.basename(file_path)
    visualize_track(lons, lats, centerline, channel1, channel2, file_name)
    
    # 保存中心线和通道数据
    output_dir = 'output'
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
    
    # 保存中心线数据
    centerline_data = pd.DataFrame({
        '经度': centerline[0],
        '纬度': centerline[1]
    })
    centerline_data.to_csv(os.path.join(output_dir, f'{os.path.splitext(file_name)[0]}_centerline.csv'), index=False)
    
    # 保存通道数据
    channel1_data = pd.DataFrame({
        '经度': channel1[0],
        '纬度': channel1[1]
    })
    channel1_data.to_csv(os.path.join(output_dir, f'{os.path.splitext(file_name)[0]}_channel1.csv'), index=False)
    
    channel2_data = pd.DataFrame({
        '经度': channel2[0],
        '纬度': channel2[1]
    })
    channel2_data.to_csv(os.path.join(output_dir, f'{os.path.splitext(file_name)[0]}_channel2.csv'), index=False)
    
    print(f"文件处理完成: {file_path}")

# 处理目录中的所有文件
def process_directory(directory):
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.endswith('.xls') or file.endswith('.xlsx'):
                file_path = os.path.join(root, file)
                process_file(file_path)

if __name__ == '__main__':
    # 处理所有文件
    base_dir = 'd:\\工作文件\\小船数据\\新建文件夹\\航道数据收集'
    process_directory(base_dir)
    print("所有文件处理完成！")

File: 数据/航道数据收集/track_centerline.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <windows.h>

// 定义点结构体
typedef struct {
    double lon;
    double lat;
} Point;

// 定义轨迹结构体
typedef struct {
    Point* points;
    int count;
} Track;

// 读取CSV文件（假设Excel已转换为CSV）
Track* read_csv_file(const char* file_path) {
    FILE* fp = fopen(file_path, "r");
    if (!fp) {      
        printf("无法打开文件: %s\n", file_path);
        return NULL;
    }

    // 计算文件中的行数
    int line_count = 0;
    char buffer[256];
    while (fgets(buffer, sizeof(buffer), fp)) {
        line_count++;
    }
    rewind(fp);

    // 分配内存
    Track* track = (Track*)malloc(sizeof(Track));
    track->points = (Point*)malloc(sizeof(Point) * line_count);
    track->count = 0;

    // 读取数据
    char* token;
    while (fgets(buffer, sizeof(buffer), fp)) {
        token = strtok(buffer, ",");
        if (token) {
            // 假设第一列为经度，第二列为纬度
            track->points[track->count].lon = atof(token);
            token = strtok(NULL, ",");
            if (token) {
                track->points[track->count].lat = atof(token);
                track->count++;
            }
        }
    }

    fclose(fp);
    return track;
}

// 三次样条插值结构体
typedef struct {
    double* x;
    double* y;
    double* y2;
    int n;
} Spline;

// 三次样条插值初始化
Spline* spline_init(double* x, double* y, int n) {
    Spline* spline = (Spline*)malloc(sizeof(Spline));
    spline->x = x;
    spline->y = y;
    spline->n = n;
    spline->y2 = (double*)malloc(sizeof(double) * n);

    // 计算二阶导数
    double* u = (double*)malloc(sizeof(double) * (n-1));
    spline->y2[0] = 0.0;
    u[0] = 0.0;

    for (int i = 1; i < n-1; i++) {
        double p = (x[i] - x[i-1]) / (x[i+1] - x[i-1]);
        double q = p * spline->y2[i-1] + 2.0;
        spline->y2[i] = (p - 1.0) / q;
        u[i] = (y[i+1] - y[i]) / (x[i+1] - x[i]) - (y[i] - y[i-1]) / (x[i] - x[i-1]);
        u[i] = (6.0 * u[i] / (x[i+1] - x[i-1]) - p * u[i-1]) / q;
    }

    spline->y2[n-1] = 0.0;
    for (int i = n-2; i >= 0; i--) {
        spline->y2[i] = spline->y2[i] * spline->y2[i+1] + u[i];
    }

    free(u);
    return spline;
}

// 三次样条插值计算
double spline_eval(Spline* spline, double x) {
    int i, j, k;
    double h, b, a;

    // 二分查找
    i = 0;
    j = spline->n - 1;
    while (j - i > 1) {
        k = (i + j) / 2;
        if (spline->x[k] > x) {
            j = k;
        } else {
            i = k;
        }
    }

    h = spline->x[j] - spline->x[i];
    a = (spline->x[j] - x) / h;
    b = (x - spline->x[i]) / h;

    return a * spline->y[i] + b * spline->y[j] + 
           ((a*a*a - a) * spline->y2[i] + (b*b*b - b) * spline->y2[j]) * h*h / 6.0;
}

// 计算轨迹中心线
Track* calculate_centerline(Track* track) {
    if (track->count < 2) {
        return track;
    }

    // 创建参数t
    int n = track->count;
    double* t = (double*)malloc(sizeof(double) * n);
    double* lons = (double*)malloc(sizeof(double) * n);
    double* lats = (double*)malloc(sizeof(double) * n);

    for (int i = 0; i < n; i++) {
        t[i] = (double)i / (n-1);
        lons[i] = track->points[i].lon;
        lats[i] = track->points[i].lat;
    }

    // 初始化样条插值
    Spline* spline_lon = spline_init(t, lons, n);
    Spline* spline_lat = spline_init(t, lats, n);

    // 生成新的点
    int new_count = 100;
    Track* centerline = (Track*)malloc(sizeof(Track));
    centerline->points = (Point*)malloc(sizeof(Point) * new_count);
    centerline->count = new_count;

    for (int i = 0; i < new_count; i++) {
        double t_new = (double)i / (new_count-1);
        centerline->points[i].lon = spline_eval(spline_lon, t_new);
        centerline->points[i].lat = spline_eval(spline_lat, t_new);
    }

    // 释放内存
    free(t);
    free(lons);
    free(lats);
    free(spline_lon->y2);
    free(spline_lon);
    free(spline_lat->y2);
    free(spline_lat);

    return centerline;
}

// 将轨迹分成上下两个通道
void split_into_channels(Track* centerline, Track* channel1, Track* channel2) {
    if (centerline->count < 2) {
        return;
    }

    // 计算轨迹方向
    double sum_dx = 0.0, sum_dy = 0.0;
    for (int i = 0; i < centerline->count - 1; i++) {
        sum_dx += centerline->points[i+1].lon - centerline->points[i].lon;
        sum_dy += centerline->points[i+1].lat - centerline->points[i].lat;
    }

    double mean_angle = atan2(sum_dy, sum_dx);
    double perpendicular_angle1 = mean_angle + M_PI / 2.0;
    double perpendicular_angle2 = mean_angle - M_PI / 2.0;

    // 偏移距离（约10米）
    double offset = 0.0001;

    // 分配内存
    channel1->points = (Point*)malloc(sizeof(Point) * centerline->count);
    channel1->count = centerline->count;
    channel2->points = (Point*)malloc(sizeof(Point) * centerline->count);
    channel2->count = centerline->count;

    // 计算通道点
    for (int i = 0; i < centerline->count; i++) {
        channel1->points[i].lon = centerline->points[i].lon + offset * cos(perpendicular_angle1);
        channel1->points[i].lat = centerline->points[i].lat + offset * sin(perpendicular_angle1);
        channel2->points[i].lon = centerline->points[i].lon + offset * cos(perpendicular_angle2);
        channel2->points[i].lat = centerline->points[i].lat + offset * sin(perpendicular_angle2);
    }
}

// 保存轨迹数据到CSV文件
void save_track_to_csv(Track* track, const char* file_path) {
    FILE* fp = fopen(file_path, "w");
    if (!fp) {
        printf("无法创建文件: %s\n", file_path);
        return;
    }

    fprintf(fp, "经度,纬度\n");
    for (int i = 0; i < track->count; i++) {
        fprintf(fp, "%lf,%lf\n", track->points[i].lon, track->points[i].lat);
    }

    fclose(fp);
}

// 处理单个文件
void process_file(const char* file_path) {
    printf("处理文件: %s\n", file_path);

    // 读取CSV文件（假设Excel已转换为CSV）
    Track* track = read_csv_file(file_path);
    if (!track) {
        return;
    }

    // 计算中心线
    Track* centerline = calculate_centerline(track);

    // 分成上下通道
    Track channel1, channel2;
    split_into_channels(centerline, &channel1, &channel2);

    // 生成输出文件名
    char output_dir[] = "output";
    CreateDirectoryA(output_dir, NULL);

    char base_name[256];
    char* ext = strrchr(file_path, '.');
    if (ext) {
        strncpy(base_name, file_path, ext - file_path);
        base_name[ext - file_path] = '\0';
    } else {
        strcpy(base_name, file_path);
    }

    char centerline_file[512];
    char channel1_file[512];
    char channel2_file[512];

    sprintf(centerline_file, "%s\\%s_centerline.csv", output_dir, strrchr(base_name, '\\') ? strrchr(base_name, '\\') + 1 : base_name);
    sprintf(channel1_file, "%s\\%s_channel1.csv", output_dir, strrchr(base_name, '\\') ? strrchr(base_name, '\\') + 1 : base_name);
    sprintf(channel2_file, "%s\\%s_channel2.csv", output_dir, strrchr(base_name, '\\') ? strrchr(base_name, '\\') + 1 : base_name);

    // 保存数据
    save_track_to_csv(centerline, centerline_file);
    save_track_to_csv(&channel1, channel1_file);
    save_track_to_csv(&channel2, channel2_file);

    // 释放内存
    free(track->points);
    free(track);
    free(centerline->points);
    free(centerline);
    free(channel1.points);
    free(channel2.points);

    printf("文件处理完成: %s\n", file_path);
}

// 遍历目录处理所有文件
void process_directory(const char* directory) {
    char search_path[512];
    sprintf(search_path, "%s\\*.csv", directory);

    WIN32_FIND_DATAA find_data;
    HANDLE hFind = FindFirstFileA(search_path, &find_data);

    if (hFind != INVALID_HANDLE_VALUE) {
        do {
            if (!(find_data.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)) {
                char file_path[512];
                sprintf(file_path, "%s\\%s", directory, find_data.cFileName);
                process_file(file_path);
            }
        } while (FindNextFileA(hFind, &find_data));
        FindClose(hFind);
    }

    // 处理子目录
    sprintf(search_path, "%s\\*", directory);
    hFind = FindFirstFileA(search_path, &find_data);

    if (hFind != INVALID_HANDLE_VALUE) {
        do {
            if ((find_data.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) && 
                strcmp(find_data.cFileName, ".") != 0 && 
                strcmp(find_data.cFileName, "..") != 0) {
                char subdir_path[512];
                sprintf(subdir_path, "%s\\%s", directory, find_data.cFileName);
                process_directory(subdir_path);
            }
        } while (FindNextFileA(hFind, &find_data));
        FindClose(hFind);
    }
}

int main() {
    // 处理所有CSV文件
    const char* base_dir = "d:\\工作文件\\小船数据\\新建文件夹\\航道数据收集";
    process_directory(base_dir);
    printf("所有文件处理完成！\n");
    return 0;
}